Signal/src/ViewControllers/CropScaleImageViewController.swift

```diff
@@ -0,0 +1,501 @@
+                // Since the translation state reflects the "upper left" corner of the crop region, we need to
+                // adjust the translation when scaling to preserve the "center" of the crop region.
+                srcTranslation.x += (srcCropSizeBeforeScalePoints.width - srcCropSizeAfterScalePoints.width) * 0.5
+                srcTranslation.y += (srcCropSizeBeforeScalePoints.height - srcCropSizeAfterScalePoints.height) * 0.5
+
+                // Update translation.
+                let viewSizePoints = imageView.frame.size
+                let srcCropSizePoints = CGSize(width:srcDefaultCropSizePoints.width / imageScale,
+                                               height:srcDefaultCropSizePoints.height / imageScale)
+
+                let viewToSrcRatio = srcCropSizePoints.width / viewSizePoints.width
+
+                let gestureTranslation = CGPoint(x:location.x - lastPinchLocation.x,
+                                                 y:location.y - lastPinchLocation.y)
+
+                srcTranslation = CGPoint(x:srcTranslation.x + gestureTranslation.x * -viewToSrcRatio,
+                                         y:srcTranslation.y + gestureTranslation.y * -viewToSrcRatio)
+
+                lastPinchLocation = location
+                lastPinchScale = sender.scale
+            }
+            break
+        case .cancelled, .failed:
+            srcTranslation = srcTranslationAtPinchStart
+            imageScale = imageScaleAtPinchStart
+            break
+        }
+
+        updateImageLayout()
+    }
+
+    var srcTranslationAtPanStart: CGPoint = CGPoint.zero
+
+    func handlePan(sender: UIPanGestureRecognizer) {
+        switch (sender.state) {
+        case .possible:
+            break
+        case .began:
+            srcTranslationAtPanStart = srcTranslation
+            break
+        case .changed, .ended:
+            guard let imageView = self.imageView else {
+                return
+            }
+            let viewSizePoints = imageView.frame.size
+            let srcCropSizePoints = CGSize(width:srcDefaultCropSizePoints.width / imageScale,
+                                           height:srcDefaultCropSizePoints.height / imageScale)
+
+            let viewToSrcRatio = srcCropSizePoints.width / viewSizePoints.width
+
+            let gestureTranslation =
+                sender.translation(in: sender.view)
+
+            // Update translation.
+            srcTranslation = CGPoint(x:srcTranslationAtPanStart.x + gestureTranslation.x * -viewToSrcRatio,
+                                     y:srcTranslationAtPanStart.y + gestureTranslation.y * -viewToSrcRatio)
+            break
+        case .cancelled, .failed:
+            srcTranslation
+                = srcTranslationAtPanStart
+            break
+        }
+
+        updateImageLayout()
+    }
+
+    func handleDoubleTap(sender: UIPanGestureRecognizer) {
+        if (sender.state == .recognized) {
```

Comment:
TIOLI:

1. I was expecting this to zoom in.
2. iMessage avatar editing doesn't seem to do anything with a double tap.

Reply:
okay, I'll skip double tap.